### PR TITLE
chore: upgrade all dependencies, devDependencies

### PR DESCRIPTION
**Pull request by dhruvdutt**
**What kind of change does this PR introduce?**
Feature/Upgrade

[Dependency Status](https://david-dm.org/dhruvdutt/webpack-cli) [Dependency Status](https://david-dm.org/dhruvdutt/webpack-cli?type=dev)

**Did you add tests for your changes?**
Updated migrate snapshot (https://github.com/webpack/webpack-cli/pull/290/commits/031b798d54cbb2b3209035f3f2ec6416f08882fb)

**Summary**
- Upgrades all dependencies, devDependencies to latest except `yargs`*. :rocket:
- Update [`supports-color`](https://github.com/chalk/supports-color) usage to resolve breaking changes
- Remove unused devDependencies: `flow-bin`, `flow-remove-types` 
- **Drops support for Node.js v4, going forward we only support v6+ (same as webpack-dev-server >= [3.0.0](https://github.com/webpack/webpack-dev-server/releases/tag/v3.0.0) and webpack >= [4.0.0](https://github.com/webpack/webpack/releases/tag/v4.0.0))**
    - Other dependencies that dropped Node.js v4 support:
        - cross-spawn >= [6.0.0](https://github.com/moxystudio/node-cross-spawn/blob/master/CHANGELOG.md#breaking-changes) 
        - inquirer >= [4.0.0](https://github.com/SBoudrias/Inquirer.js/releases/tag/v4.0.0)

*We will need to doublecheck on `yargs` changelog for breaking changes in version [10.0.0](https://github.com/yargs/yargs/blob/master/CHANGELOG.md#breaking-changes-1) and [11.0.0](https://github.com/yargs/yargs/blob/master/CHANGELOG.md#breaking-changes)

**Does this PR introduce a breaking change?**
No

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -123,7 +123,6 @@
     "cz-customizable": "^5.2.0",
     "eslint": "^4.2.0",
     "eslint-plugin-node": "^6.0.1",
-    "flow-bin": "^0.66.0",
     "flow-remove-types": "^1.2.1",
```
Comment (ematipico) [inline package.json]: Sorry to bother again! Just noticed that we can also remove `flow-remove-types`!

Comment (dhruvdutt) [reply package.json]: Ah! No worries. I just ran [`npm-check`](https://www.npmjs.com/package/npm-check) to check for all unused deps and this is the list of unused deps returned in output:

Can you please confirm which ones to keep/remove?

`global`
`recast`
`uglifyjs-webpack-plugin`
`@commitlint/cli`
`@commitlint/prompt-cli`
`commitizen`
`conventional-changelog-lint-config-cz`
~`flow-remove-types`~
`husky`
`lint-staged`
`schema-utils`

**Comment by dhruvdutt** [comment]
Not sure why it fails. I've already rebased and my [package.json](https://github.com/dhruvdutt/webpack-cli/blob/upgrade/deps/package.json#L36-L41) doesn't have the part that was causing the CI to fail.

File: package.json (hunk 1 above)
Comment (ev1stensberg) [reply package.json]: No one here

Comment (dhruvdutt) [reply package.json]: I've cleaned `flow-bin`, `flow-remove-types` as suggested by @ematipico 
Also, updated PR description,

**Comment by dhruvdutt** [comment]
@ev1stensberg Reverted back to `"yargs": "9.0.1"` and added a note in PR description.

File: bin/webpack.js [hunk 2]
```diff
@@ -318,7 +318,7 @@
 			});
 
 			if (typeof outputOptions.colors === "undefined")
-				outputOptions.colors = require("supports-color");
+				outputOptions.colors = require("supports-color").stdout;
```
Comment (ev1stensberg) [inline bin/webpack.js]: Anything changed in a newer version?

Comment (dhruvdutt) [reply bin/webpack.js]: Doesn't look like anything changed relevant to us other than the API.
There are no changelogs or release notes but you can take a look at this PR [#64](https://github.com/chalk/supports-color/pull/64).

Comment (ev1stensberg) [reply bin/webpack.js]: Remove the stdout and you're good. Doesn't work with it

**Comment by ev1stensberg** [comment]
rebase against master

**Comment by dhruvdutt** [comment]
Already rebased. There's some weird issue that ANSI terminal color codes aren't parsed due to some reason when the tests run. Check the screenshot below. Printing `stdout` like:
```js
console.log(stdout[5]);
console.log(stdout);
```
image

However this doesn't happen when we add `stdout` in `supports-color` usage.
```js
require("supports-color").stdout
```
image

There are no changelogs for `supports-color` versions but we can go through the [latest docs](https://github.com/chalk/supports-color#usage). 

Should we add `stdout` back to `supports-color` usage?

**Comment by dhruvdutt** [comment]
Already rebased.

All tests basically fail with this error: (Check expected and contain values)

image

It should match but it doesn't. Most likely there's some weird issue that ANSI terminal color codes aren't parsed due to some reason when the tests run. Check the screenshot below. Printing `stdout` like:
```js
console.log(stdout[5]);
console.log(stdout);
```
image

However this doesn't happen when we add `stdout` in `supports-color` usage.
```js
require("supports-color").stdout
```
image

There are no changelogs for `supports-color` versions but we can go through the [latest docs](https://github.com/chalk/supports-color#usage). 

Currently added `stdout` back to `supports-color`.
LMK if we have some other fix or missing something.

**Comment by ev1stensberg** [comment]
Conflicts

**Comment by dhruvdutt** [comment]
@ev1stensberg Resolved. :100:

File: bin/process-options.js [hunk 3]
```diff
@@ -37,7 +37,7 @@ module.exports = function processOptions(yargs, argv) {
 	});
 
 	if (typeof outputOptions.colors === "undefined")
-		outputOptions.colors = require("supports-color");
+		outputOptions.colors = require("supports-color").stdout;
```
Comment (ev1stensberg) [inline bin/process-options.js]: No stdout here

File: bin/webpack.js (hunk 2 above)
Comment (ev1stensberg) [inline bin/webpack.js]: No stdout here

File: bin/webpack.js [hunk 4]
```diff
@@ -80,7 +80,7 @@
 			type: "boolean",
 			alias: "colors",
 			default: function supportsColor() {
-				return require("supports-color");
+				return require("supports-color").stdout;
```
Comment (ev1stensberg) [inline bin/webpack.js]: no stdout

File: bin/webpack.js (hunk 2 above)
Comment (dhruvdutt) [reply bin/webpack.js]: Removing this breaks the tests. Did you read my [above comment](https://github.com/webpack/webpack-cli/pull/290#issuecomment-369672078)?
Do we want the tests to fail?

Comment (ev1stensberg) [reply bin/webpack.js]: I want colors to work as intended. If the tests are failing you gotta fix em 👩‍💻✌️

Comment (dhruvdutt) [reply bin/webpack.js]: Coolio. So, in that case we can use something like [ansi-parser] (https://www.npmjs.com/package/ansi-parser) to parse `stdout` inside tests. WDYT?

Comment (ev1stensberg) [reply bin/webpack.js]: Remove the stdout and lemme have a look

Comment (dhruvdutt) [reply bin/webpack.js]: Done. You're up captain. :guardsman:

**Pull request merged**